hive-metastore/src/main/java/org/apache/iceberg/hive/HiveClientPool.java

```diff
@@ -83,6 +83,20 @@ protected HiveMetaStoreClient reconnect(HiveMetaStoreClient client) {
     return client;
   }
 
+  @Override
+  protected boolean failureDetection(HiveMetaStoreClient client, Exception ex) {
+    if (null != ex && ex instanceof MetaException) {
+      try {
+        client.getDatabase("default");
```

Comment:
I would prefer to have a configurable timeout for this, so we do not end up doubling every failed metastore call. MetaException can be returned even if the HMS is working correctly. Also it would be good to be able to turn off this check with the same config.

One more thing to consider is the time of the duplicated call. Would it be possible to return the original result and do the check asynchronously? Or in case of failure we always retry now?

If I remember correctly theoretically you can habe a HMS without a "default" db.
I remember that there was an API call to get the HMS version which should be good, or if it does not exist in  Hive 2.3 then maybe `get_current_notificationEventId` would be good for the quick health check.

What do you think?

Reply:
1. "I would prefer to have a configurable timeout for this",  This is worth it.
2. "MetaException can be returned even if the HMS is working correctly",  For this, the implement does not affect the normal exception `MetaException `  thrown, because when the method returns false, the exception is thrown by `ClientPool.run`.
3. "it would be good to be able to turn off this check with the same config",  for this, i don't think it's necessary, because query exception is not acceptable for users. So we need to achieve automatic recovery, is not it?
4. “If I remember correctly theoretically you can have a HMS without a`default`db.”， for this problem, the implement  only return false when the exception is not the `TTransportException`,  whether the operation `client.getDatabase("default")`  is OK or not.

Reply:
2. For example: When you try to run `client.getTable(null, "tableName")` you will receive MetaException. You can take a look at the test cases [here](https://github.com/apache/hive/tree/master/standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/client) and see that there are multiple cases where MetaException is thrown. In this case MetaException is an acceptable result returning a valid error, and we do not need to try to reconnect
3. Based on 2. we might get a MetaException even if everything is working correctly
4. Collecting db information is more costly than returning the only row from a table. We need a check method which is as cheap as possible.

So basically I think we can not rule out receiving the MetaException when we are working correctly. We might decide that in relation to Iceberg we think the risk is minimal enough to accept the risk of double calls and the risk worth the code simplicity. How complicated would be to create a "theoretically" sound solution?

Checked the `client.getServerVersion()` client method. This might be less costly than calling `client.getDatabase("default")`.